Update Billing documentation (#1015)
bullet-train-co/bullet_train-core: bullet_train/docs/billing/stripe.md

end
```

### A.3. Bundle Install

```
bundle install
```

**Proceed to the [Installation Instructions](#installation-instructions) section below**

## Option B: Using the Open Source version of Bullet Train Billing via git

### B.1. Add the Git Repositories

```
# Below the comment labelled YOUR GEMS in your Gemfile
gem "bullet_train-billing", git: "https://github.com/bullet-train-pro/bullet_train-billing.git"
gem "bullet_train-billing-stripe", git: "https://github.com/bullet-train-pro/bullet_train-billing-stripe.git"
```

> ### B.3. Bundle Install

Reviewer: ```suggestion
### B.2. Bundle Install
```

> gem "bullet_train-billing", git: "https://github.com/bullet-train-pro/bullet_train-billing.git", ref: "The git reference you selected"
> gem "bullet_train-billing-stripe", git: "https://github.com/bullet-train-pro/bullet_train-billing-stripe.git", ref: "The git reference you selected"

Reviewer: ```suggestion
gem "bullet_train-billing", git: "https://github.com/bullet-train-pro/bullet_train-billing.git"
gem "bullet_train-billing-stripe", git: "https://github.com/bullet-train-pro/bullet_train-billing-stripe.git"
```

end
```

### A.3. Bundle Install

```
bundle install
```

**Proceed to the [Installation Instructions](#installation-instructions) section below**

## Option B: Using the Open Source version of Bullet Train Billing via git

> - [Git history for bullet_train-billing](https://github.com/bullet-train-pro/bullet_train-billing/commits/main/)
> - [Git history for bullet_train-billing-stripe](https://github.com/bullet-train-pro/bullet_train-billing-stripe/commits/main/)
> 
> ### B.2. Add the Git Repositories

Reviewer: ```suggestion
### B.1. Add the Git Repositories
```

end
```

### A.3. Bundle Install

```
bundle install
```

**Proceed to the [Installation Instructions](#installation-instructions) section below**

## Option B: Using the Open Source version of Bullet Train Billing via git

> ### B.1. Find a Compatible Version
> 
> It is possible you will need to find a release that is compatible with your version of BulletTrain. As of the time of writing,
> `bullet_train-billing` and `bullet_train-billing-stripe` do not have versions tagged in git. You will need to go through the commit
> history of each project and identify an appropriate git reference.

Reviewer: ```suggestion
```

end
```

### A.3. Bundle Install

```
bundle install
```

> **Proceed to the [Installation Instructions](#installation-instructions) section below

Reviewer: @jiblits, just making sure, is this a note to yourself to add the `**` at the end of the line?

## Option B: Using the Open Source version of Bullet Train Billing via git

### B.1. Find a Compatible Version

It is possible you will need to find a release that is compatible with your version of BulletTrain. As of the time of writing,
`bullet_train-billing` and `bullet_train-billing-stripe` do not have versions tagged in git. You will need to go through the commit

> history of each project and identify an appropriate git reference.

Reviewer: Hopefully we'll get those repos merged into `core` before too terribly long and this will sort itself out. That being said, I don't _expect_ that there's really incompatible versions if you're reasonably up to date on both the starter repo and the billing gems, so I'm not sure that this bit is necessary. (If there are, I'd love to know some details.)

> history of each project and identify an appropriate git reference.

Reviewer: Have you run into incompatibility problems on this? If there aren't explicit known problems then I'd rather remove this paragraph to keep things simple. If there are known problems we should document them as clearly as possible.

- [Git history for bullet_train-billing](https://github.com/bullet-train-pro/bullet_train-billing/commits/main/)
- [Git history for bullet_train-billing-stripe](https://github.com/bullet-train-pro/bullet_train-billing-stripe/commits/main/)

For this example we will be using `bullet_train-billing` version 1.0.12 and `bullet_train-billing-stripe` version 1.0.4.

### B.2. Add the Git Repositories

```
# Below the comment labelled YOUR GEMS in your Gemfile
gem "bullet_train-billing", git: "https://github.com/bullet-train-pro/bullet_train-billing.git", ref: "dd05d76"
gem "bullet_train-billing-stripe", git: "https://github.com/bullet-train-pro/bullet_train-billing-stripe.git", ref: "b1e3c2e"

> ```

Reviewer: I think for the main example we probably don't want to include a `ref`, just because people will probably copy that as-is and end up getting "pinned" to whatever that `ref` points to.

Author: I'll swap it out with "a version appropriate git ref"

Reviewer: For simplicity I'd rather leave the `ref` out all together so that people can copy and paste from the docs and have it Just Work. I think this is introducing a level of complexity that most people just don't need to grapple with.

#### 5.2. Enable Stripe Webhooks

  - Visit [https://dashboard.stripe.com/test/webhooks/create](https://dashboard.stripe.com/test/webhooks/create).
  - Use the default "add an endpoint" form.
  - Set "endpoint URL" to `https://YOUR-SUBDOMAIN.ngrok.io/webhooks/incoming/stripe_webhooks`.
  - Under "select events to listen to" choose "select all events" and click "add events".
  - Finalize the creation of the endpoint by clicking "add endpoint".

*Note:* Depending on your region and when your Stripe account was created, you may be presented with a different admin interface.
For more details on creating webhooks, please refer to the [Stripe Workbench documentation](https://docs.stripe.com/workbench#webhooks)

#### 5.3. Configure Stripe Webhooks Signing Secret

After creating the webhook endpoint, click "reveal" under the heading "signing secret". Copy the `whsec_...` value into your `config/application.yml` like so:

```yaml
STRIPE_WEBHOOKS_ENDPOINT_SECRET: whsec_aBunchOfRandomCharacters
```

#### 5.4. Test Sample Webhook Delivery

 - Restart your Rails server with `rails restart`.

>  - Trigger a test webhook using the Stripe CLI. Such as `stripe trigger payment_intent.succeeded`

Reviewer: ```suggestion
 - Trigger a test webhook using the Stripe CLI (for instance `stripe trigger payment_intent.succeeded`). You just want to make sure that it results in an HTTP status code of 201.
```

#### 5.2. Enable Stripe Webhooks

  - Visit [https://dashboard.stripe.com/test/webhooks/create](https://dashboard.stripe.com/test/webhooks/create).
  - Use the default "add an endpoint" form.
  - Set "endpoint URL" to `https://YOUR-SUBDOMAIN.ngrok.io/webhooks/incoming/stripe_webhooks`.
  - Under "select events to listen to" choose "select all events" and click "add events".
  - Finalize the creation of the endpoint by clicking "add endpoint".

*Note:* Depending on your region and when your Stripe account was created, you may be presented with a different admin interface.
For more details on creating webhooks, please refer to the [Stripe Workbench documentation](https://docs.stripe.com/workbench#webhooks)

#### 5.3. Configure Stripe Webhooks Signing Secret

After creating the webhook endpoint, click "reveal" under the heading "signing secret". Copy the `whsec_...` value into your `config/application.yml` like so:

```yaml
STRIPE_WEBHOOKS_ENDPOINT_SECRET: whsec_aBunchOfRandomCharacters
```

#### 5.4. Test Sample Webhook Delivery

 - Restart your Rails server with `rails restart`.

>  - Trigger a test webhook using the Stripe CLI. Such as `stripe trigger payment_intent.succeeded`

Reviewer: We probably want to leave the bit that mentions that we're just looking for a 201 status code. Just to make it clear that we don't expect anything else to happen.

### 6. Test Creating a Subscription

> Bullet Train comes preconfigured with a "freemium" plan, which will keep new and existing account working as normal. A new **billing** menu item will appear in the **Team** menu.

Reviewer: ```suggestion
Bullet Train comes preconfigured with a "freemium" plan, which will keep new and existing accounts working as normal. A new **billing** menu item will appear in the **Team** menu.
```

### 2.4. Run Migrations

### 2. Run Migrations

> 

Reviewer: This seems to duplicate the `### 2.4 Run Migrations` heading just above it. We probably don't need both?

Author: Whoops! Nope, `### 2.4 Run Migrations` should be removed.

```
rake db:migrate
```

### 3. Configure Your Products

Bullet Train defines subscription plans and pricing options in `config/models/billing/products.yml` and defines the translatable elements of these plans in `config/locales/en/billing/products.en.yml`. We recommend just getting started with these plans to ensure your setup is working before customizing the attributes of these plans.

### 4. Configure Stripe

#### 4.1. Create API Keys with Stripe

  - Create a Stripe account if you don't already have one.
  - Visit [https://dashboard.stripe.com/test/apikeys](https://dashboard.stripe.com/test/apikeys).
  - Create a new secret key.

**Note:** By default we're linking to the "test mode" page for API keys so you can get up and running in development. When you're ready to deploy to production, you'll have to repeat this step and toggle the "test mode" option off to provision real API keys for live payments.

#### 4.2. Configure Stripe API Keys Locally

Edit `config/application.yml` and add your new Stripe secret key to the file:

```yaml

> STRIPE_SECRET_KEY: sk_test_aBunchOfRandomCharacters

Author: I figured swapping these keys out with very obviously not valid keys. The keys have probably been rotated out or were never valid though I figured making this change would be OK. Also added in the `_test_` portion since that's what I was seeing from the Stripe API when I was following the procedure.

STRIPE_PRODUCT_BASIC_MONTHLY_2025_PRICE_ID: price_1QpbQSKKAAAAAAAAAAAAAAAA
STRIPE_PRODUCT_BASIC_ANNUAL_2025_PRICE_ID: price_1QpbQSKKBBBBBBBBBBBBBBBB
STRIPE_PRODUCT_PRO_ANNUAL_2025_PRICE_ID: price_1QpbQSKKCCCCCCCCCCCCCCCC
```

### 5. Wire Up Webhooks

Basic subscription creation will work without receiving and processing Stripe's webhooks. However, advanced payment workflows like SCA payments and customer portal cancelations and plan changes require receiving webhooks and processing them.

#### 5.1. Ensure HTTP Tunneling is Enabled

Although Stripe provides free tooling for receiving webhooks in your local environment, the officially supported mechanism for doing so in Bullet Train is using [HTTP Tunneling with ngrok](/docs/tunneling.md). This is because we provide support for many types of webhooks across different platforms and packages, so we already need to have ngrok in play.

Ensure you've completed the steps from [HTTP Tunneling with ngrok](/docs/tunneling.md), including updating `BASE_URL` in `config/application.yml` and restarting your Rails server.

#### 5.2. Enable Stripe Webhooks

>   - Visit [https://dashboard.stripe.com/test/workbench/webhooks](https://dashboard.stripe.com/test/workbench/webhooks).

Reviewer: For me `https://dashboard.stripe.com/test/workbench/webhooks` isn't a valid address and it redirects to `https://dashboard.stripe.com/test/dashboard`. I can access the webhook overview screen at `https://dashboard.stripe.com/test/webhooks`.

>   - Click the "+ Add destination" button

Reviewer: For me the button says "+ Add endpoint".

  - In the "Select events" section choose the following:
    - Events from: "Your account"
    - API version: Your current version
    - Events: Click the checkbox "Select all events"

>       - **This raises a warning about latency and performance. To discuss during documentation review to determine if this can be trimmed down to more specific events**

Reviewer: This is a good question.

  - Click "Continue ->"
  - Select the "Webhook endpoint" option and click "Continue ->"
  - Set the "Endpoint URL" to `https://YOUR-SUBDOMAIN.ngrok.io/webhooks/incoming/stripe_webhooks`
  - Set the Description to "Bullet Train Development Webhook Endpoint"

>   - Click "Create destination"

Author: This reflects the experience I had when setting up bullet train to my Stripe account. I don't know if this is the case for everyone though.

Just some context for my experience here:

- My account was created ~1 week ago
- I'm Canadian. So I don't know if there's a different UI based on region or not. I haven't worked with Stripe that much

Reviewer: It does seem like things are different either for newer accounts, or for Canadian accounts. This is what I see after clicking "+ Add endpoint".

![CleanShot 2025-02-07 at 11 35 42](https://github.com/user-attachments/assets/4a811746-16e6-40c8-b00a-8bfd03a9cd6e)

Author: What I'll do is restore it to what was there previously and add a note that these instructions may not exactly reflect every developers setup experience.

This "products" list is what Stripe will display to users as upgrade and downgrade options in the customer portal. You shouldn't list any products here that aren't properly configured in your Rails app, otherwise the resulting webhook will fail to process. If you want to stop offering a plan, you should remove it from this list as well.

### 8. Finalize Webhooks Testing by Managing a Subscription

In the same account where you created your first test subscription, go into the "billing" menu and click "manage" on that subscription. This will take you to the Stripe Billing customer portal.

Once you're in the customer portal, you should test upgrading, downgrading, and canceling your subscription and clicking "⬅ Return to {Your Application Name}" in between each step to ensure that each change you're making is properly reflected in your Bullet Train application. This will let you know that webhooks are being properly delivered and processed and all the products in both systems are properly mapped in both directions.

#### 8.1 Understanding the Product configuration yml

> Your `products.yml` should have a structure similar to the following:

Author: The product.yml has a couple of portions to it that I wasn't certain I understood. I was thinking calling out an example of what it could look like and explaining the configuration variables could be handy for someone getting set up with the billing product.

Reviewer: Sounds like a good idea.

```yaml
free:
  limits:
    memberships:
      count: 1
      enforcement: hard
    scaffolding/absolutely_abstract/creative_concepts:
      count: 0
      enforcement: hard
basic:
  price: basic_2021
  image: "streamline-icon-idea-message.png"
  limits:
    memberships:
      count: 3
      enforcement: hard
    scaffolding/absolutely_abstract/creative_concepts:
      count: null
  prices:
    basic_2021:
      amount: 1000
      currency: USD
      duration: 1
      interval: month
      quantity: memberships
    basic_2021:
      amount: 10000
      currency: USD
      duration: 1
      interval: month
      quantity: memberships
```

Each root level entry (i.e. `free` and `basic`) are the types of subscriptions that will be available to your customers. Each subscription has the following:

>   - `price`: The product price key **discuss during PR to get a better understanding of what this means here**

Author: Want to call this one out specifically as I'm _completely_ guessing what this is.

Reviewer: It's basically a unique identifier that will identify this price. I usually use something like `basic_2025` to help identify when that price became active.

  - `image`: The product image from `assets/images/products` that is presented to the buyer from the new subscription page
  - `limits`: The resource constraints on what a user can do on this subscription. To learn more, refer to [Bullet Train Usage Limits](/docs/billing/usage.md)
  - `prices`: The list of pricing options for this product, where each price will have the following:
    - `amount`: The cost of the product in *cents*
    - `currency`: The currency that the customer will be billed in (i.e. EUR for Euros, CAD for Canadian Dollars)
    - `duration`: Used for subscriptions to define how many units the subscription is valid for

>     - `interval`: The unit of time of the subscription (i.e. day, month, year)

Reviewer: We might mention that `duration` and `interval` are related. For instance if you set `duration: 2` and `interval: month` then people will be billed every other month. If you set `duration: 3` and `interval: month` then they'll be billed every third month.

>     - `quantity`: The relation used to determine how many of this product are being ordered.

Author: Are there more details I can link to or include to help explain how the quantity is calculated? From what I was seeing in development it was based on the count of how many people were on my team.

In the case of a fixed-price can this be absent or should it be `1`?

Reviewer: Yes, that's exactly right. If you see `quantity` to the value `memberships` then the subscription will be configured for however many users you have on your team. Setting it to `1` (or any other number) will make it a flat fee that is `quantity * amount`.

      - If your pricing model charges per seat, this should be set to `memberships`

>       - If you are charging a fix rate, this field should be omitted

Reviewer: Not omitted, set to `1`.

    - `highlight`: When set to true, this product will *pop out* from the other product options that are presented to the user

For full details on product pricing, please refer to the [Product and Prices Overview Documentation on Stripe](https://docs.stripe.com/products-prices/overview)

#### 8.2 Replacing Test Products with your own

  - Before making any changes, be sure to cancel any subscriptions

>     - **NOTE FOR THE REVIEWER**: Is this the right approach? How should people approach removing old products with still-active (even if cancelled) subscriptions? 

Reviewer: This is a good question, I'll have to look into it.

  - Go to https://dashboard.stripe.com/test/products and find the test products that were added
  - Click on the "..." menu to the right and click "Archive product"
  - Remove the product keys from `config/application.yml`

>   - Replace the boilerplate product entries in `config/models/billing/products.yml`

Author: Are there any caveats that need to be listed for developers who are done with the test products that come from the base installation?

Reviewer: I think the main caveat is that the `free` plan is kind of a special one and should not be renamed. (I think... 😬)

#### 8.3 Customizing the Pricing Page content

The pricing page and pricing page translations are located within the `bullet_train-billing` gem. If you need to customize either of those you will need to eject them. For more details on ejecting components from Bullet Train gems, refer to the section on [Indirection](https://bullettrain.co/docs/indirection).

**Ejecting the pricing page**

```
bin/resolve bullet_train-billing/app/views/account/billing/subscriptions/new.html.erb --eject
```

**Ejecting the subscriptions translations**

**NOTE FOR THE REVIEWER**: I am not certain on how to eject the translations files from the gem, the following does not work

```
bin/resolve en.account.billing.subscriptions.new.pricing.header --eject

> ```

Author: I couldn't get this to work. I was able to get other keys to resolve from the core project but not from the pro gems.

I'm guessing I'm doing something wrong and it's possible to get this working. As a fallback, in case this isn't working I added some quick curl instructions. Though, if we can get the file ejecting properly I'd rather use that to reduce the risk of drift between the actual file and what is expected here in these.

Reviewer: I'll look into this and see if we can improve the situation.
